src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -2958,8 +3000,14 @@ ConnectionStateMachine(WorkerSession *session)
 
 			case MULTI_CONNECTION_CONNECTED:
 			{
-				/* connection is ready, run the transaction state machine */
-				TransactionStateMachine(session);
+				/*
+				 * Connection is ready, and we have unfinished tasks.
+				 * So, run the transaction state machine.
+				 */
+				if (execution->unfinishedTaskCount > 0)
```

Comment:
It seems like we should be able to skip TransactionStateMachine as soon as the worker has 0 unfinished tasks. Do we have an easy way to calculate it?

Reply:
Seems so